ksql-common/src/main/java/io/confluent/ksql/schema/ksql/KsqlSchema.java

```diff
@@ -40,39 +53,251 @@
 @Immutable
 public final class KsqlSchema {
 
-  private final Schema schema;
+  private static final Consumer<Schema> NO_ADDITIONAL_VALIDATION = schema -> {
+  };
+
+  private static final Map<Type, Consumer<Schema>> VALIDATORS =
+      ImmutableMap.<Type, Consumer<Schema>>builder()
+          .put(Type.BOOLEAN, NO_ADDITIONAL_VALIDATION)
+          .put(Type.INT32, NO_ADDITIONAL_VALIDATION)
+          .put(Type.INT64, NO_ADDITIONAL_VALIDATION)
+          .put(Type.FLOAT64, NO_ADDITIONAL_VALIDATION)
+          .put(Type.STRING, NO_ADDITIONAL_VALIDATION)
+          .put(Type.ARRAY, KsqlSchema::validateArray)
+          .put(Type.MAP, KsqlSchema::validateMap)
+          .put(Type.STRUCT, KsqlSchema::validateStruct)
+          .build();
+
+  private final ConnectSchema schema;
 
   public static KsqlSchema of(final Schema schema) {
     return new KsqlSchema(schema);
   }
 
   private KsqlSchema(final Schema schema) {
-    this.schema = requireNonNull(schema, "schema").schema();
-    throwIfMutable(this.schema);
+    this.schema = validate(Objects.requireNonNull(schema, "schema"), true);
```

Comment:
why not have this call `build()`?

Reply:
Because its not building anything ;) It's validating! One of the things it validates is that its not mutable, i.e. that it can be cast to `ConnectSchema`.